File: matlab/examples/transmission_lines/Finite_Stripline.m
% example demonstrating the use of a stripline terminated by a resistance
% (c) 2013 Thorsten Liebig

close all
clear
clc

%% setup the simulation %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
physical_constants;
unit = 1e-6; % specify everything in um
SL_length = 50000;
SL_width = 520;
SL_height = 500;
substrate_thickness = SL_height;
substrate_epr = 3.66;
f_max = 7e9;

Air_Spacer = 20000;

%% setup FDTD parameters & excitation function %%%%%%%%%%%%%%%%%%%%%%%%%%%%
FDTD = InitFDTD();
FDTD = SetGaussExcite( FDTD, f_max/2, f_max/2 );
BC   = {'MUR' 'MUR' 'MUR' 'MUR' 'MUR' 'MUR'};
FDTD = SetBoundaryCond( FDTD, BC );

%% setup CSXCAD geometry & mesh %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
CSX = InitCSX();
resolution = c0/(f_max*sqrt(substrate_epr))/unit /50; % resolution of lambda/50
mesh.x = SmoothMeshLines( [-SL_length/2 0 SL_length/2], resolution, 1.5 ,0 );
mesh.y = SmoothMeshLines( [0 SL_width/2+[-resolution/3 +resolution/3*2]/4], resolution/4 , 1.5 ,0);
mesh.y = SmoothMeshLines( [-10*SL_width -mesh.y mesh.y 10*SL_width], resolution, 1.3 ,0);
mesh.z = linspace(0,substrate_thickness,5);
mesh.z = sort(unique([mesh.z -mesh.z]));

%% substrate
CSX = AddMaterial( CSX, 'RO4350B' );
CSX = SetMaterialProperty( CSX, 'RO4350B', 'Epsilon', substrate_epr );
start = [mesh.x(1),   mesh.y(1),   mesh.z(1)];
stop  = [mesh.x(end), mesh.y(end), mesh.z(end)];
CSX = AddBox( CSX, 'RO4350B', 0, start, stop );

%% add air spacer
mesh.x = [mesh.x mesh.x(1)-Air_Spacer mesh.x(end)+Air_Spacer];
mesh.y = [mesh.y mesh.y(1)-Air_Spacer mesh.y(end)+Air_Spacer];
mesh.z = [mesh.z mesh.z(1)-Air_Spacer mesh.z(end)+Air_Spacer];
mesh = SmoothMesh(mesh, c0/f_max/unit/20);
CSX = DefineRectGrid( CSX, unit, mesh );

%% SL port
CSX = AddMetal( CSX, 'PEC' );
portstart = [ -SL_length/2, -SL_width/2, 0];
portstop  = [ 0,         SL_width/2, 0];
[CSX,port{1}] = AddStripLinePort( CSX, 999, 1, 'PEC', portstart, portstop, SL_height, 'x', [0 0 -1], 'ExcitePort', true, 'Feed_R', 50, 'MeasPlaneShift',  SL_length/3);

portstart = [+SL_length/2, -SL_width/2, 0];
portstop  = [0          ,  SL_width/2, 0];
[CSX,port{2}] = AddStripLinePort( CSX, 999, 2, 'PEC', portstart, portstop, SL_height, 'x', [0 0 -1], 'MeasPlaneShift',  SL_length/3, 'Feed_R', 50);

% bottom PEC plane
CSX = AddBox(CSX, 'PEC', 999, [-SL_length/2 -10*SL_width -SL_height],[+SL_length/2 +10*SL_width -SL_height]);
% top PEC plane
CSX = AddBox(CSX, 'PEC', 999, [-SL_length/2 -10*SL_width SL_height],[+SL_length/2 +10*SL_width SL_height]);

%% write/show/run the openEMS compatible xml-file
Sim_Path = ['tmp_' mfilename];
Sim_CSX = 'stripline.xml';

[status, message, messageid] = rmdir( Sim_Path, 's' ); % clear previous directory
[status, message, messageid] = mkdir( Sim_Path ); % create empty simulation folder

WriteOpenEMS( [Sim_Path '/' Sim_CSX], FDTD, CSX );
CSXGeomPlot( [Sim_Path '/' Sim_CSX] );
RunOpenEMS( Sim_Path, Sim_CSX );

%% post-processing
close all
f = linspace( 1e6, f_max, 1601 );
port = calcPort( port, Sim_Path, f, 'RefImpedance', 50);

s11 = port{1}.uf.ref./ port{1}.uf.inc;
s21 = port{2}.uf.ref./ port{1}.uf.inc;

plot(f/1e9,20*log10(abs(s11)),'k-','LineWidth',2);
hold on;
grid on;
plot(f/1e9,20*log10(abs(s21)),'r--','LineWidth',2);
legend('S_{11}','S_{21}');
ylabel('S-Parameter (dB)','FontSize',12);
xlabel('frequency (GHz) \rightarrow','FontSize',12);
ylim([-50 2]);

File: python/Tests/Finite_Stripline.py
# -*- coding: utf-8 -*-
"""
 Finite Stripline Test (MUR-terminated, resistance-terminated ports)

 Verifies StripLinePort with MUR absorbing boundaries and explicit top/bottom
 PEC ground planes.  Complements Stripline.py which uses PML termination.

 Pass criteria:
   max(dB(S11)) < -30 dB   (low reflection; MUR is less accurate than PML)
   min(dB(S21)) > -0.5 dB  (near-lossless transmission)
   max(dB(S21)) < +0.01 dB (sign-error guard)

 Tested with
  - python 3.14
  - openEMS v0.0.36+

 (c) 2026 Thorsten Liebig <thorsten.liebig@gmx.de>

"""

import os, tempfile
import numpy as np

from CSXCAD  import ContinuousStructure
from openEMS import openEMS
from openEMS.physical_constants import *
from openEMS.ports import StripLinePort

### Setup the simulation
Sim_Path = os.path.join(tempfile.gettempdir(), 'Finite_Stripline')

unit          = 1e-6  # drawing unit in um
SL_length     = 50000
SL_width      = 520
SL_height     = 500   # distance from strip to each ground plane
substrate_epr = 3.66
f_max         = 7e9
Air_Spacer    = 20000

### Setup FDTD parameters & excitation
FDTD = openEMS(EndCriteria=1e-4)
FDTD.SetGaussExcite(f_max/2, f_max/2)
FDTD.SetBoundaryCond(['MUR']*6)

### Setup CSXCAD geometry & mesh
CSX = ContinuousStructure()
FDTD.SetCSX(CSX)
mesh = CSX.GetGrid()
mesh.SetDeltaUnit(unit)

resolution = C0 / (f_max * np.sqrt(substrate_epr)) / unit / 50
third_mesh = np.array([-1/3, 2/3]) * resolution / 4

mesh.AddLine('x', [-SL_length/2, 0, SL_length/2])
mesh.SmoothMeshLines('x', resolution, ratio=1.5)

mesh.AddLine('y', [0])
mesh.AddLine('y', SL_width/2 + third_mesh)
mesh.SmoothMeshLines('y', resolution/4, ratio=1.5)
y_pos = mesh.GetLines('y')
mesh.AddLine('y', np.concatenate([-y_pos, [-10*SL_width, 10*SL_width]]))
mesh.SmoothMeshLines('y', resolution, ratio=1.3)

mesh.AddLine('z', np.linspace(0, SL_height, 5))
mesh.AddLine('z', -np.linspace(0, SL_height, 5))

# Air spacer around the domain for MUR boundaries
x_lines = mesh.GetLines('x')
y_lines = mesh.GetLines('y')
z_lines = mesh.GetLines('z')
mesh.AddLine('x', [x_lines[0] - Air_Spacer, x_lines[-1] + Air_Spacer])
mesh.AddLine('y', [y_lines[0] - Air_Spacer, y_lines[-1] + Air_Spacer])
mesh.AddLine('z', [z_lines[0] - Air_Spacer, z_lines[-1] + Air_Spacer])
mesh.SmoothMeshLines('all', C0 / f_max / unit / 20)

### Substrate
substrate = CSX.AddMaterial('RO4350B', epsilon=substrate_epr)
substrate.AddBox([-SL_length/2, -10*SL_width, -SL_height],
                 [ SL_length/2,  10*SL_width,  SL_height])

### Stripline ports (include the metal strip)
pec = CSX.AddMetal('PEC')

portstart = [-SL_length/2, -SL_width/2, 0]
portstop  = [0,             SL_width/2, 0]
port1 = StripLinePort(CSX, 1, pec, portstart, portstop, 'x', 'z', SL_height,
                      excite=1, priority=999,
                      MeasPlaneShift=SL_length/3, Feed_R=50)

portstart = [SL_length/2, -SL_width/2, 0]
portstop  = [0,            SL_width/2, 0]
port2 = StripLinePort(CSX, 2, pec, portstart, portstop, 'x', 'z', SL_height,
                      priority=999, MeasPlaneShift=SL_length/3, Feed_R=50)

ports = [port1, port2]

### Explicit top and bottom PEC ground planes
gnd = CSX.AddMetal('GND')
gnd.AddBox([-SL_length/2, -10*SL_width, -SL_height],
           [ SL_length/2,  10*SL_width, -SL_height], priority=999)
gnd.AddBox([-SL_length/2, -10*SL_width,  SL_height],
           [ SL_length/2,  10*SL_width,  SL_height], priority=999)


if 0:  # debugging only
    CSX_file = os.path.join(Sim_Path, 'fin_stripline.xml')
    if not os.path.exists(Sim_Path):
        os.mkdir(Sim_Path)
    CSX.Write2XML(CSX_file)
    from CSXCAD import AppCSXCAD_BIN
    os.system(AppCSXCAD_BIN + ' "{}"'.format(CSX_file))

### Run the simulation
FDTD.Run(Sim_Path, cleanup=True, exact_endcriteria=True)

### Post-processing
f = np.linspace(1e6, f_max, 1601)
for port in ports:
    port.CalcPort(Sim_Path, f, ref_impedance=50)

s11 = ports[0].uf_ref / ports[0].uf_inc
s21 = ports[1].uf_ref / ports[0].uf_inc

s11_dB = 20 * np.log10(np.abs(s11))
s21_dB = 20 * np.log10(np.abs(s21))

### Pass / fail checks
mask = f > 100e6
print(f'max(dB(S11)) = {np.max(s11_dB[mask]):.1f} dB')
print(f'min(dB(S21)) = {np.min(s21_dB[mask]):.1f} dB,  max(dB(S21)) = {np.max(s21_dB[mask]):.2f} dB')

assert np.max(s11_dB[mask]) < -30, \
    f'FAIL: max(dB(S11)) = {np.max(s11_dB[mask]):.1f} dB, expected < -30 dB'
assert np.min(s21_dB[mask]) > -0.5, \
    f'FAIL: min(dB(S21)) = {np.min(s21_dB[mask]):.1f} dB, expected > -0.5 dB'
assert np.max(s21_dB[mask]) < 0.05, \
    f'FAIL: max(dB(S21)) = {np.max(s21_dB[mask]):.2f} dB, expected < +0.05 dB (sign error?)'

print('PASS')

if 0:  # set to 1 for debugging plots
    import matplotlib.pyplot as plt

    fig, axis = plt.subplots(num='S-Parameters', tight_layout=True)
    axis.plot(f/1e9, s11_dB, 'k-',  linewidth=2, label='$S_{11}$')
    axis.plot(f/1e9, s21_dB, 'r--', linewidth=2, label='$S_{21}$')
    axis.grid()
    axis.set_xmargin(0)
    axis.set_ylim([-50, 2])
    axis.set_xlabel('Frequency (GHz)')
    axis.set_ylabel('S-Parameter (dB)')
    axis.legend()

    plt.show()
